File: engine/src/core/utils/UUID.cpp
#include "Engine.h"
#include "UUID.h"


namespace Paper
{
	static std::random_device randomDevice;
	static std::mt19937_64 engine(randomDevice());
	static std::uniform_int_distribution<uint64_t> distribution;

	UUID::UUID()
		: uuid(distribution(engine))
	{
	}

	UUID::UUID(uint64_t uuid)
		: uuid(uuid)
	{
	}

	UUID::UUID(std::string id)
	{
		std::stringstream ss;
		ss << std::hex << id;
		ss >> uuid;
	}

	void UUID::Set(std::string id)
	{
		std::stringstream ss;
		ss << std::hex << id;
		ss >> uuid;
	}

	void UUID::Set(uint64_t id)
	{
		uuid = id;
	}

	uint64_t UUID::toUInt64() const
	{
		return operator uint64_t();
	}

	std::string UUID::toString() const
	{
		return operator std::string();
	}
}

File: scriptcore/Source/Paper/Core/UUID.cs
using System;

namespace Paper
{
    public class UUID
    {
        private ulong _uuid;

        public UUID(ulong _UUID)
        {
            _uuid = _UUID;
        }

        public UUID(string _UUID)
        {
            _uuid = Convert.ToUInt64(_UUID, 16);
        }

        public UUID(UUID _UUID)
        {
            this._uuid = _UUID._uuid;
        }

        public void Set(ulong _UUID)
        {
            _uuid = _UUID;
        }

        public void Set(string _UUID)
        {
            _uuid = Convert.ToUInt64(_UUID, 16);
        }

        public override string ToString()
        {
            return _uuid.ToString("x16");
        }

        public ulong ToUlong()
        {
            return _uuid;
        }

        public static implicit operator ulong(UUID _UUID)
        {
            return _UUID._uuid;
        }

        public static implicit operator UUID(ulong _UUID)
        {
            return new UUID(_UUID);
        }

        public static implicit operator string(UUID _UUID)
        {
            return _UUID.ToString();
        }

        public static implicit operator UUID(string _UUID)
        {
            return new UUID(_UUID);
        }
    }
}